src/Native/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -1537,7 +1537,12 @@ bool GCToOSInterface::VirtualCommit(void* address, size_t size)
 //  true if it has succeeded, false if it has failed
 bool GCToOSInterface::VirtualDecommit(void* address, size_t size)
 {
-    return mprotect(address, size, PROT_NONE) == 0;
+    // TODO: This can fail, however the GC does not handle the failure gracefully
+    // Explicitly calling mmap instead of mprotect here makes it
+    // that much more clear to the operating system that we no
+    // longer need these pages. Also, GC depends on re-commited pages to
+    // be zeroed-out.
+    return mmap(address, size, PROT_NONE, MAP_FIXED | MAP_ANON | MAP_PRIVATE, -1, 0) != NULL;
```

Comment:
I wasn't aware that you could overwrite existing mappings with `mmap`, so I did some reading and the language in macOS and Linux documentation seems to indicate that it'll remove any existing mappings that exist at the target location. Do you know if that means that if we e.g. decommit a subset of a full map (e.g. end of a segment), the kernel will toss out the entire mapping, or does it replace the mapping only at the target address + size leaving the rest of the mapping intact?

Reply:
It will replace the mapping only at the target address + size. This fix does exactly what we are doing in CoreCLR PAL today: https://github.com/dotnet/coreclr/blob/0ee3b5e64a98dc71aefed2304fe4bcf7f66ca9f5/src/pal/src/map/virtual.cpp#L1512

Reply:
good to know! thanks!